### PR TITLE
feat(*/grpc): update composed behaviour

### PR DESCRIPTION
**Pull request by Freyskeyd**
https://img.shields.io/badge/3%20PR-grpc%20over%20p2p-blue 

# Description
Previous PR: #332 
Next PR: #321 

This PR is activating the gRPC behaviour on the p2p composed behaviour in order to start using it in the synchronizer

## PR Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-tce-synchronizer/src/lib.rs [hunk 1]
```diff
@@ -78,6 +92,104 @@ pub enum SynchronizerError {
...
+impl GrpcSynchronizerService for SynchronizerService {
+    async fn fetch_certificates(
+        &self,
+        request: Request<FetchCertificatesRequest>,
+    ) -> Result<Response<FetchCertificatesResponse>, Status> {
+        let request = request.into_inner();
+        let certificate_ids: Vec<CertificateId> = request
+            .certificates
+            .into_iter()
+            .map(|c| c.try_into().unwrap())
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/lib.rs:117]: Feels like we shouldn't panic here (didn't we fix a bug in cert id parsing last week?)

File: crates/topos-tce-synchronizer/src/lib.rs [hunk 2]
```diff
@@ -78,6 +92,104 @@ pub enum SynchronizerError {
...
+            .collect();
+
+        let response =
+            if let Ok(certs) = self.validator_store.get_certificates(&certificate_ids[..]) {
+                let certs: Vec<_> = certs
+                    .into_iter()
+                    .filter_map(|v| v.map(|c| c.certificate.try_into().unwrap()))
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/lib.rs:124]: Same here: why is it ok to panic here?

**Comment by Freyskeyd** [comment]
@dvdplm The PR is still in draft, code will change a bit I guess

File: crates/topos-tce-synchronizer/src/checkpoints_collector/tests.rs [hunk 3]
```diff
@@ -42,9 +53,74 @@ fn encode() {
...
+    let res = client
+        .send_request::<_, FetchCertificatesResponse>(
+            boot_node.keypair.public().to_peer_id(),
+            req,
+            RetryPolicy::NoRetry,
+            SYNCHRONIZER_PROTOCOL,
+        )
+        .await;
+
+    println!("RECV gRPC over P2P response: {:?}", res);
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/checkpoints_collector/tests.rs:114]:
```suggestion
    debug!("RECV gRPC over P2P response: {:?}", res);
```

File: crates/topos-p2p/src/lib.rs [hunk 4]
```diff
@@ -26,6 +26,7 @@ pub use libp2p::Multiaddr;
 pub use libp2p::PeerId;
 pub use runtime::Runtime;
 
+pub(crate) mod temp_grpc;
```
Comment (dvdplm) [inline crates/topos-p2p/src/lib.rs:29]: Is this module used?

File: crates/topos-p2p/src/lib.rs [hunk 5]
```diff
@@ -35,7 +36,60 @@ pub const TOPOS_ECHO: &str = "topos_echo";
...
+    pub struct GrpcOverP2P {
+        pub(crate) proxy_sender: mpsc::Sender<Command>,
+    }
+
+    impl GrpcOverP2P {
+        pub async fn create<S: GrpcClient>(
+            &self,
+            peer: PeerId,
+        ) -> Result<S::Output, tonic::transport::Error> {
+            warn!("Creating new instance of GRPC CLIENT FOR P2P");
```
Comment (dvdplm) [inline crates/topos-p2p/src/lib.rs:71]:
```suggestion
            debug!("Creating new instance of GRPC CLIENT FOR P2P");
```

File: crates/topos-p2p/src/lib.rs [hunk 6]
```diff
@@ -35,7 +36,60 @@ pub const TOPOS_ECHO: &str = "topos_echo";
...
+                    id,
+                    response: sender,
+                })
+                .await;
+
+            let connection = recv.await.unwrap();
+
+            let connected = connection.into_future().await.unwrap();
+
+            info!("Connected: {:?}", connected);
```
Comment (dvdplm) [inline crates/topos-p2p/src/lib.rs:88]:
```suggestion
            debug!("Connected: {:?}", connected);
```

File: crates/topos-p2p/src/lib.rs [hunk 7]
```diff
@@ -35,7 +36,60 @@ pub const TOPOS_ECHO: &str = "topos_echo";
...
+                .send(Command::NewProxiedQuery {
+                    peer,
+                    id,
+                    response: sender,
+                })
+                .await;
+
+            let connection = recv.await.unwrap();
+
+            let connected = connection.into_future().await.unwrap();
```
Comment (dvdplm) [inline crates/topos-p2p/src/lib.rs:86]: Why is it ok to panic here?

File: crates/topos-p2p/src/error.rs [hunk 8]
```diff
@@ -38,6 +38,9 @@ pub enum P2PError {
 
     #[error("Unable to execute shutdown on the p2p runtime: {0}")]
     ShutdownCommunication(mpsc::error::SendError<oneshot::Sender<()>>),
+
+    #[error("Unable to create gRPC client")]
+    UnableToCreateGrpcClient,
```
Comment (dvdplm) [inline crates/topos-p2p/src/error.rs:43]: Would be good to include the tonic error here. As it is it's not obvious what the problem was?

File: crates/topos-p2p/src/constant.rs [hunk 9]
```diff
@@ -39,3 +39,6 @@ pub const GRPC_P2P_TOPOS_PROTOCOL: &str = "/topos-grpc-p2p/1.0";
 
 // FIXME: Considered as constant until customizable and exposed properly in the genesis file
 pub const TCE_BOOTNODE_PORT: u16 = 9090;
+
```
Comment (dvdplm) [inline crates/topos-p2p/src/constant.rs:42]: Total nitpick but this file should be named `constants` (plural).

File: crates/topos-p2p/src/command.rs [hunk 10]
```diff
@@ -62,6 +63,11 @@ pub enum Command {
         topic: &'static str,
         data: Vec<u8>,
     },
+    NewProxiedQuery {
```
Comment (dvdplm) [inline crates/topos-p2p/src/command.rs:66]: Docs would be good

File: crates/topos-p2p/src/runtime/mod.rs [hunk 11]
```diff
@@ -33,6 +33,7 @@ use tokio::sync::{
     oneshot::{Receiver, Sender},
 };
 use tokio_stream::StreamExt;
+use tonic::transport::server::Routes;
```
Comment (dvdplm) [inline crates/topos-p2p/src/runtime/mod.rs:36]: Not used I think?

File: crates/topos-p2p/src/runtime/handle_event.rs [hunk 12]
```diff
@@ -83,6 +86,7 @@ impl
                 self.active_listeners.insert(listener_id);
             }
             SwarmEvent::OutgoingConnectionError { peer_id, error, .. } => {
+                info!("OutgoingConnectionError");
```
Comment (dvdplm) [inline crates/topos-p2p/src/runtime/handle_event.rs:89]:
```suggestion
```

File: crates/topos-p2p/src/runtime/handle_event.rs [hunk 13]
```diff
@@ -115,11 +119,11 @@ impl
             }
 
             incoming_connection_error @ SwarmEvent::IncomingConnectionError { .. } => {
-                debug!("{:?}", incoming_connection_error);
+                error!("{:?}", incoming_connection_error);
             }
 
             SwarmEvent::IncomingConnection { local_addr, .. } => {
-                debug!("IncomingConnection {local_addr}")
+                info!("IncomingConnection {local_addr}")
```
Comment (dvdplm) [inline crates/topos-p2p/src/runtime/handle_event.rs:126]: debug perhaps?

File: crates/topos-p2p/src/runtime/handle_event/grpc.rs [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+use tracing::info;
+
+use crate::{behaviour::grpc, Runtime};
+
+use super::EventHandler;
+
+#[async_trait::async_trait]
+impl EventHandler<grpc::Event> for Runtime {
+    async fn handle(&mut self, event: grpc::Event) {
+        info!("gRPC event: {:?}", event);
```
Comment (dvdplm) [inline crates/topos-p2p/src/runtime/handle_event/grpc.rs:17]: Hmm, this isn't doing much! What do we need this for? Is it a placeholder for other code?

**Comment by dvdplm** [comment]
Still a draft but here's a review anyway. :)

File: crates/topos-p2p/src/constants.rs [hunk 15]
```diff
@@ -34,3 +34,6 @@ pub const GRPC_P2P_TOPOS_PROTOCOL: &str = "/topos-grpc-p2p/1.0";
 
 // FIXME: Considered as constant until customizable and exposed properly in the genesis file
 pub const TCE_BOOTNODE_PORT: u16 = 9090;
+
+/// Swarm idle connection timeout
+pub const IDLE_CONNECTION_TIMEOUT_SECONDS: u64 = 30;
```
Comment (hadjiszs) [inline crates/topos-p2p/src/constants.rs:39]: nit:

```suggestion
pub const IDLE_CONNECTION_TIMEOUT = Duration::from_secs(30);
```

File: crates/topos-tce-synchronizer/src/checkpoints_collector/tests.rs [hunk 16]
```diff
@@ -42,9 +53,73 @@ fn encode() {
     assert_eq!(y, req);
 }
 
-#[test]
-fn sync_same_certificate_twice() {
-    //
+#[rstest]
+#[test_log::test(tokio::test)]
+#[timeout(Duration::from_secs(10))]
+async fn network_test() {
```
Comment (hadjiszs) [inline crates/topos-tce-synchronizer/src/checkpoints_collector/tests.rs:59]: I'd be more explicit in the likes of `check_fetch_certificates_with_synchronizer` (or something else)

**Review by hadjiszs — APPROVED** [review]

**Pull request merged**